File: notification-service/src/main/java/org/tggc/notificationservice/config/KafkaProducerConfig.java
package org.tggc.notificationservice.config;

import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;
import org.springframework.kafka.support.serializer.JsonSerializer;

import java.util.Map;

import static org.apache.kafka.clients.producer.ProducerConfig.BOOTSTRAP_SERVERS_CONFIG;
import static org.apache.kafka.clients.producer.ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG;
import static org.apache.kafka.clients.producer.ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG;

@Configuration
public class KafkaProducerConfig {
    @Value("${spring.kafka.bootstrap-servers}")
    private String bootstrapServers;

    private Map<String, Object> producerProperties() {
        return Map.of(
                BOOTSTRAP_SERVERS_CONFIG, bootstrapServers,
                KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class,
                VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class
        );
    }

    @Bean
    public ProducerFactory<String, Object> producerFactory() {
        return new DefaultKafkaProducerFactory<>(producerProperties());
    }

    @Bean
    public KafkaTemplate<String, Object> kafkaTemplate(ProducerFactory<String, Object> factory) {
        return new KafkaTemplate<>(factory);
    }
}

File: event-service/src/main/java/org/tggc/eventservice/config/kafka/KafkaProducerConfig.kt
package org.tggc.eventservice.config.kafka

import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import org.springframework.kafka.support.serializer.JsonSerializer
import java.util.Map

@Configuration
open class KafkaProducerConfig(
    @param:Value($$"${spring.kafka.bootstrap-servers}")
    private val bootstrapServers: String
) {

    private fun producerProperties(): MutableMap<String?, Any?> {
        return Map.of<String?, Any?>(
            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers,
            ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java,
            ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer::class.java
        )
    }

    @Bean
    open fun producerFactory(): ProducerFactory<String?, Any?> {
        return DefaultKafkaProducerFactory<String?, Any?>(producerProperties())
    }

    @Bean
    open fun kafkaTemplate(factory: ProducerFactory<String?, Any?>): KafkaTemplate<String?, Any?> {
        return KafkaTemplate<String?, Any?>(factory)
    }
}
